Rebus.MongoDb2/MongoDbTimeoutStorage.cs

```diff
@@ -57,18 +82,35 @@ public void Add(Timeout.Timeout newTimeout)
         /// </summary>
         public DueTimeoutsResult GetDueTimeouts()
         {
-            var result = collection.Find(Builders<BsonDocument>.Filter.Lte(TimeProperty, RebusTimeMachine.Now()))
-                                   .Sort(Builders<BsonDocument>.Sort.Ascending(TimeProperty));
-
-			return new DueTimeoutsResult(result.Project(r =>
-					new DueMongoTimeout(GetString(r, ReplyToProperty),
-						GetString(r, CorrIdProperty),
-						r[TimeProperty].ToUniversalTime(),
-						GetGuid(r, SagaIdProperty),
-						GetString(r, DataProperty),
-						collection,
-						(ObjectId) r[IdProperty])
-					).ToList());
+			var dueTimeouts = new List<DueTimeout>();
+			var currentDate = RebusTimeMachine.Now();
+            var dueTimeoutFilter = Builders<BsonDocument>.Filter.Lte(TimeProperty,currentDate);
+            var unlockedTimeoutFilter = Builders<BsonDocument>.Filter.Eq(DueLockProperty, BsonNull.Value) |
+                                Builders<BsonDocument>.Filter.Lt(DueLockProperty, currentDate);
+
+            var timeoutReadyToBeFiredFilter = dueTimeoutFilter & unlockedTimeoutFilter;
+
+            var dueLockShifted = currentDate + _lockTimeoutsOffset;
+
+            for (var i = 0; i < _maxDueTimeoutsRetrieved; i++)
+            {
+                var findOneAndUpdateOptions = new FindOneAndUpdateOptions<BsonDocument> { Sort = Builders<BsonDocument>.Sort.Ascending(TimeProperty).Ascending(DueLockProperty) };
+                var r = collection.FindOneAndUpdate(timeoutReadyToBeFiredFilter, Builders<BsonDocument>.Update.Set(DueLockProperty, dueLockShifted), findOneAndUpdateOptions);
+                if (r != null)
```

Comment:
Avoid loopipng the whole cycle, and do a break once you get the first null.

Reply:
Ok, just wanted the loop to be as greedy as posible to give it an instance the chance to retrieve as much due timeouts as possible.